src/persist/src/operators/stream.rs

```diff
@@ -702,19 +709,73 @@ where
-                                .write(&[anti_update])
-                                .await
-                                .expect("error persisting retraction");
+                            to_retract.push(anti_update);
                             continue;
                         }
                         session.give(update);
                     }
+
+                    let write_future = write.write(&to_retract);
+                    to_retract.clear();
+
+                    // We are not using the capability for the main output later, but we are
+                    // holding on to it to keep the frontier from advancing because that frontier
+                    // is used downstream to track how far we have persisted. This is used, for
+                    // example, by upsert()/persist()/seal()/conditional_seal() operators
+                    // and await_frontier().
+                    pending_futures.push_back((
+                        cap.delayed(cap.time()),
+                        cap.retain_for_output(error_output_port),
+                        write_future,
+                    ));
+                });
+
+                // Swing through all pending futures and see if they're ready. Ready futures will
+                // invoke the Activator, which will make sure that we arrive here, even when there
+                // are no changes in the input frontier or new input.
+                let waker = futures_util::task::waker_ref(&activator);
+                let mut context = Context::from_waker(&waker);
+
+                while let Some((cap, error_cap, pending_future)) = pending_futures.front_mut() {
+                    match Pin::new(pending_future).poll(&mut context) {
+                        std::task::Poll::Ready(result) => {
+                            match result {
+                                Ok(seq_no) => {
+                                    log::trace!(
+                                        "In {}, finished writing for time: {}, seq_no: {:?}",
+                                        &operator_name,
+                                        cap.time(),
+                                        seq_no,
+                                    );
+                                }
+                                Err(e) => {
+                                    let mut session = error_output.session(&error_cap);
+                                    let error = format!(
+                                        "In {}, error writing data for time {}: {}",
+                                        &operator_name,
+                                        error_cap.time(),
+                                        e
+                                    );
+                                    log::error!("{}", error);
+
+                                    // TODO: make error retractable? Probably not...
+                                    session.give((error, *error_cap.time(), 1));
```

Comment:
if one of these fails, we need to permanently wedge the frontier from advancing, right? otherwise, we could see incorrect duplicates on restart

Reply:
Yes, which ties in with your [other comment](https://github.com/MaterializeInc/materialize/pull/9229#discussion_r755216901), right?